tests/cuda/matrix_vector_mf.cuh

```diff
@@ -0,0 +1,118 @@
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// This is a template for matrix-vector products with the Helmholtz equation
+// (zero and first derivatives) on different kinds of meshes (Cartesian,
+// general, with and without hanging nodes).
+
+#include "../tests.h"
+
+#include <deal.II/matrix_free/cuda_matrix_free.h>
+#include <deal.II/matrix_free/cuda_fe_evaluation.cuh>
+
+#include <deal.II/lac/cuda_vector.h>
+
+template <int dim, int fe_degree, typename Number, int n_q_points_1d>
+class HelmholtzOperatorQuad
+{
+public :
+  __device__ void operator()(
+    CUDAWrappers::FEEvaluation<dim,fe_degree,n_q_points_1d,1,Number> *fe_eval,
+    const unsigned int q_point) const;
+};
+
+
+
+template <int dim, int fe_degree, typename Number, int n_q_points_1d>
+__device__ void
+HelmholtzOperatorQuad<dim,fe_degree,Number,n_q_points_1d>::operator()(
+  CUDAWrappers::FEEvaluation<dim,fe_degree,n_q_points_1d,1,Number> *fe_eval,
+  const unsigned int q) const
+{
+  fe_eval->submit_value(Number(10)*fe_eval->get_value(q),q);
+  fe_eval->submit_gradient(fe_eval->get_gradient(q),q);
+}
+
+
+
+template <int dim, int fe_degree, typename Number, int n_q_points_1d>
+class HelmholtzOperator
+{
+public:
+  __device__ void operator()(
+    const unsigned int cell,
+    const typename CUDAWrappers::MatrixFree<dim,Number>::Data *gpu_data,
+    CUDAWrappers::SharedData<dim,Number>                *shared_data,
+    const Number                                        *src,
+    Number                                              *dst) const;
+
+  static const unsigned int n_dofs_1d = fe_degree+1;
+  static const unsigned int n_local_dofs =
+    dealii::Utilities::fixed_int_power<fe_degree+1,dim>::value;
+  static const unsigned int n_q_points =
+    dealii::Utilities::fixed_int_power<n_q_points_1d,dim>::value;
+};
+
+
+
+template <int dim, int fe_degree, typename Number, int n_q_points_1d>
+__device__ void
+HelmholtzOperator<dim,fe_degree,Number,n_q_points_1d>::operator()(
+  const unsigned int                                        cell,
+  const typename CUDAWrappers::MatrixFree<dim,Number>::Data *gpu_data,
+  CUDAWrappers::SharedData<dim,Number>                *shared_data,
+  const Number                                        *src,
+  Number                                              *dst) const
+{
+  CUDAWrappers::FEEvaluation<dim,fe_degree,n_q_points_1d,1,Number> fe_eval(
+    cell, gpu_data, shared_data);
+  fe_eval.read_dof_values(src);
+  fe_eval.evaluate(true, true);
+  fe_eval.apply_quad_point_operations(
+    HelmholtzOperatorQuad<dim,fe_degree,Number,n_q_points_1d>());
+  fe_eval.integrate(true, true);
+  fe_eval.distribute_local_to_global (dst);
```

Comment:
This looks definitely like a step in the right direction. One thing I don't like is the fact that we need to expose the vectors as a plain pointer here. Even though we do use plain pointers inside the CPU version as well and could expose them, I would prefer if the handling of the pointers - which right now happens in the `cell_loop` function - could be hidden. On the other hand, I don't know if we are already ready to aim for the most appropriate design, which would be to hide away the complete function.

Just as a bit of background regarding the optimizations we could apply:
- One wants to be able to do `read_dof_values` and `evaluate` with a single call. The optimization possibilities for cell integrals are not too big, but definitely for face integrals because this allows to only read those dof values which are non-zero on a face. Similarly with `integrate` and `distributed_local_to_global`
- One might want to interleave parts of the `evaluate` and `integrate` calls with the `quad_point_operation` as most appropriate.
- We need to support multiple components (think of Taylor-Hood) with different needs on the vectors.